File: Interviews/LeetCode/Topic-Wise/Recursion/Medium/Maximum-Length-of-a-Concatenated-String-with-Unique-Characters/1239.py
class Solution:
    def helper (self, idx, s, arr) -> int:
        if idx < 0:
            return 0

        ans = self.helper (idx - 1, s, arr)
        cnt, ok = [0 for i in range(26)], True
        for c in s:
            cnt[ord(c) - 97] += 1
            ok &= cnt[ord(c) - 97] == 1
        for c in arr[idx]:
            cnt[ord(c) - 97] += 1
            ok &= cnt[ord(c) - 97] == 1
        if ok:
            ans = max(ans, self.helper (idx - 1, s + arr[idx], arr) + len(arr[idx]))
            
        return ans
        
    def maxLength (self, arr: List[str]) -> int:
        n = len(arr)
        return self.helper (n - 1, "", arr)

File: Interviews/LeetCode/Daily-Problems/2022/October/Day-24/1239.cpp
class Solution {
public:
    void helper (int idx, string s, int &ans, vector<string> &arr) {
        if (idx < 0) {
            if (s.size() > ans) {
                ans = s.size();
            }
            return;
        }

        helper (idx - 1, s, ans, arr);
        vector<int> cnt(26, 0);
        bool ok = true;
        for (char c : s) {
            cnt[c - 'a'] += 1;
            ok &= cnt[c - 'a'] == 1;
        }

        for (char c : arr[idx]) {
            cnt[c - 'a'] += 1;
            ok &= cnt[c - 'a'] == 1;
        }

        if (ok) {
            helper (idx - 1, s + arr[idx], ans, arr);
        }
    }
    int maxLength(vector<string>& arr) {
        int ans = 0, n = arr.size();
        helper (n - 1, "", ans, arr);
        return ans;
    }
};